src/colmap/geometry/homography_matrix.cc

```diff
@@ -192,41 +192,27 @@ namespace {
 
 double CheckCheiralityAndReprojErrorSum(
     const Rigid3d& cam2_from_cam1,
-    const std::vector<Eigen::Vector2d>& points1,
-    const std::vector<Eigen::Vector2d>& points2,
+    const std::vector<Eigen::Vector3d>& cam_rays1,
+    const std::vector<Eigen::Vector3d>& cam_rays2,
     std::vector<Eigen::Vector3d>* points3D) {
-  THROW_CHECK_EQ(points1.size(), points2.size());
-  const Eigen::Matrix3x4d cam1_from_world = Eigen::Matrix3x4d::Identity();
-  const Eigen::Matrix3x4d cam2_from_world = cam2_from_cam1.ToMatrix();
-  constexpr double kMinDepth = std::numeric_limits<double>::epsilon();
-  const double max_depth = 1000.0 * cam2_from_cam1.translation.norm();
+  THROW_CHECK_EQ(cam_rays1.size(), cam_rays2.size());
   double reproj_residual_sum = 0;
   points3D->clear();
-  for (size_t i = 0; i < points1.size(); ++i) {
-    Eigen::Vector3d point3D;
-    if (!TriangulatePoint(cam1_from_world,
-                          cam2_from_world,
-                          points1[i],
-                          points2[i],
-                          &point3D)) {
-      continue;
-    }
-    const Eigen::Vector3d point3D_in_cam1 =
-        cam1_from_world * point3D.homogeneous();
-    if (point3D_in_cam1.z() < kMinDepth || point3D_in_cam1.z() > max_depth) {
-      continue;
-    }
-    const Eigen::Vector3d point3D_in_cam2 =
-        cam2_from_world * point3D.homogeneous();
-    if (point3D_in_cam2.z() < kMinDepth || point3D_in_cam2.z() > max_depth) {
+  for (size_t i = 0; i < cam_rays1.size(); ++i) {
+    Eigen::Vector3d point3D_in_cam1;
+    if (!TriangulateMidPoint(
+            cam2_from_cam1, cam_rays1[i], cam_rays2[i], &point3D_in_cam1)) {
       continue;
     }
+    const Eigen::Vector3d point3D_in_cam2 = cam2_from_cam1 * point3D_in_cam1;
     const double error1 =
-        (points1[i] - point3D_in_cam1.hnormalized()).squaredNorm();
+        1 -
+        std::clamp(cam_rays1[i].dot(point3D_in_cam1.normalized()), -1.0, 1.0);
     const double error2 =
-        (points2[i] - point3D_in_cam2.hnormalized()).squaredNorm();
+        1 -
+        std::clamp(cam_rays2[i].dot(point3D_in_cam2.normalized()), -1.0, 1.0);
     reproj_residual_sum += error1 + error2;
```

Comment:
Should we update the names of this variable and of the function? This isn't a reprojection error anymore but a cosine similarity.